rsconf/component/jupyterhub.py

```diff
@@ -46,70 +45,86 @@ def internal_build_compile(self):
         )
         z.setdefault('http_timeout', 30);
         z.setdefault('template_vars', {});
-        nc = self.buildt.get_component('network')
-        z.hub_ip = nc.ip_and_net_for_host(jc.rsconf_db.host)[0]
-        z.template_d = z.run_d.join(_TEMPLATE_D)
+        self._network(z, jc)
+        self._user_lists(z)
+        z.template_d = self.__run_d.join(_TEMPLATE_D)
         z.home_d = db.user_home_path(jc, z.jupyter_run_u)
         z.cookie_secret_hex = self.secret_path_value(
-            _COOKIE_SECRET,
+            self._COOKIE_SECRET,
             gen_secret=lambda: db.random_string(length=64, is_hex=True),
             visibility='channel',
         )[0]
         z.proxy_auth_token = self.secret_path_value(
-            _PROXY_AUTH,
+            self._PROXY_AUTH,
             gen_secret=lambda: db.random_string(length=64, is_hex=True),
             visibility='channel',
         )[0]
-        # admin_users are implicitly part of whitelist
-        for x in 'admin', 'blacklist', 'whitelist':
-            y = z.get(f'{x}_users')
-            if y:
-                z[f'{x}_users_str'] = _list_to_str(y)
-        self._rsdockerspawner(jc, z)
+        self._rsdockerspawner(z)
 
     def internal_build_write(self):
         from rsconf import db
-        from rsconf import systemd
-        from rsconf.component import docker_registry
         from rsconf.component import docker
 
-        z = self.jc.jupyterhub
+        z = self.j2_ctx[self.name]
+        self._auth(z)
+        self._install_jupyterhub_db(z)
```

Comment:
@robnagler please double check these installs. I'm not positive I have it right.